cub/cub/thread/thread_reduce.cuh

```diff
@@ -43,199 +41,515 @@
+{
+  Output output;
+  static_assert(sizeof(input) == sizeof(output), "wrong size");
+  ::memcpy(&output, &input, sizeof(input));
+  return output;
+}
+
+} // namespace detail
+
 /// Internal namespace (to prevent ADL mishaps between static functions when mixing different CUB installations)
 namespace internal
 {
 
-#ifndef DOXYGEN_SHOULD_SKIP_THIS // Do not document
+/***********************************************************************************************************************
+ * Enable SIMD/Tree reduction heuristics
+ **********************************************************************************************************************/
 
 /// DPX instructions compute min, max, and sum for up to three 16 and 32-bit signed or unsigned integer parameters
 /// see DPX documetation https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#dpx
 /// NOTE: The compiler is able to automatically vectorize all cases with 3 operands
 ///       However, all other cases with per-halfword comparison need to be explicitly vectorized
-/// TODO: Remove DPX specilization when nvbug 4823237 is fixed
 ///
 /// DPX reduction is enabled if the following conditions are met:
 /// - Hopper+ architectures. DPX instructions are emulated before Hopper
 /// - The number of elements must be large enough for performance reasons (see below)
 /// - All types must be the same
 /// - Only works with integral types of 2 bytes
-/// - DPX instructions provide Min, Max, and Sum SIMD operations
+/// - DPX instructions provide Min, Max SIMD operations
 /// If the number of instructions is the same, we favor the compiler
+///
+/// length | Standard |  DPX
+///  2     |    1     |  NA
+///  3     |    1     |  NA
+///  4     |    2     |  3
+///  5     |    2     |  3
+///  6     |    3     |  3
+///  7     |    3     |  3
+///  8     |    4     |  4
+///  9     |    4     |  4
+/// 10     |    5     |  4 // ***
+/// 11     |    5     |  4 // ***
+/// 12     |    6     |  5 // ***
+/// 13     |    6     |  5 // ***
+/// 14     |    7     |  5 // ***
+/// 15     |    7     |  5 // ***
+/// 16     |    8     |  6 // ***
+
+// TODO: add Blackwell support
+
+template <typename Input, typename ReductionOp>
+_CCCL_NODISCARD _CCCL_DEVICE constexpr bool enable_generic_simd_reduction()
+{
+  using cub::detail::is_one_of;
+  using T      = ::cuda::std::__remove_cvref_t<decltype(::cuda::std::declval<Input>()[0])>;
+  using Length = ::cuda::std::integral_constant<int, cub::detail::static_size_v<Input>()>;
+  // clang-format off
+  return ((is_one_of<T, ::cuda::std::int16_t, ::cuda::std::uint16_t>() && is_one_of<ReductionOp, cub::Min, cub::Max>())
+#  if defined(_CCCL_HAS_NVFP16)
+     || (::cuda::std::is_same<T, __half>::value && is_one_of<ReductionOp, cub::Min, cub::Max, cub::Sum, cub::Mul>())
+#  endif
+#  if defined(_CCCL_HAS_NVBF16)
+     || (::cuda::std::is_same<T, __nv_bfloat16>::value &&
+         is_one_of<ReductionOp, cub::Min, cub::Max, cub::Sum, cub::Mul>())
+#  endif
```

Comment:
I find this technique really hard to read, why are we not using a struct that we specialize for half and friends?

Reply:
sorry I don't exactly understand what you are proposing. Are you concerned by the `#if defines`? Do you suggest to break this function into multiple template structs with specializations for different types?

Reply:
what do you think?
```cpp
template <typename T, typename ReductionOp>
struct enable_generic_simd_reduction_traits
{
  static constexpr bool value = cub::detail::is_one_of<T, ::cuda::std::int16_t, ::cuda::std::uint16_t>()
                             && cub::detail::is_one_of<ReductionOp, cub::Min, cub::Max>();
};

#  if defined(_CCCL_HAS_NVFP16)

template <typename ReductionOp>
struct enable_generic_simd_reduction_traits<__half, ReductionOp>
{
  static constexpr bool value = cub::detail::is_one_of<ReductionOp, cub::Min, cub::Max, cub::Sum, cub::Mul>();
};
#  endif // defined(_CCCL_HAS_NVFP16)

#  if defined(_CCCL_HAS_NVBF16)

template <typename ReductionOp>
struct enable_generic_simd_reduction_traits<__nv_bfloat16, ReductionOp>
{
  static constexpr bool value = cub::detail::is_one_of<ReductionOp, cub::Min, cub::Max, cub::Sum, cub::Mul>();
};

#  endif // defined(_CCCL_HAS_NVBF16)
```

Reply:
Yeah